### DISCUSSION TITLE
How to filter one-to-one relationships by value in the joined table?

### DISCUSSION THREAD

**Discussion by christopheraranda** [PostgREST/postgrest#3216] [Q&A] [answered]
How do you filter the results of a query of one table based on the value in a joined table that has a one-to-one relationship?

For example, given the database in the [documentation](https://postgrest.org/en/stable/references/api/resource_embedding.html#one-to-one-relationships):

```http
GET /films?select=title,technical_specs(camera) HTTP/1.1
```

how would I select only films that used the `Arriflex 35-III` camera?

**Comment by laurenceisla** [comment]
The filter works the same way as other relationships (see [Embedded Filters](https://postgrest.org/en/stable/references/api/resource_embedding.html#embedded-filters) in the docs). In this case, the query would be (`%20` is an url encoded empty space):

```http
GET /films?select=title,technical_specs(camera)&technical_specs.camera=eq.Arriflex%2035-III HTTP/1.1
```

**Reply by laurenceisla** [reply]
Now, to do a [Top-Level Filtering](https://postgrest.org/en/stable/references/api/resource_embedding.html#top-level-filtering) you'll need to add `!inner` to the query: `techinal_specs!inner(camera)`.

**Reply by christopheraranda** [reply]
Hi @laurenceisla

Thanks for the quick reply! That doesn't seem to be working for me. The joined table uses a quoted identifier for its name. So in this example, the table `technical_specs` is named `"technical-specs"`. So if I modify the example that you provided to reflect the table name:

```http
GET /films?select=title,technical-specs!inner(camera)&technical-specs.camera=eq.Arriflex%2035-III HTTP/1.1
```

The `technical-specs` are included, but all films are being returned and not just the films where the camera is `Arriflex 35-III`

**Reply by laurenceisla** [reply]
That's strange, it works for me even with the `"technical-specs"` table name. I'm using curl, and it gives me the result:

```console
$ curl "http://localhost:3000/films?select=title,technical-specs!inner(camera)&technical-specs.camera=eq.Arriflex%2035-III"
[{"title":"A movie","technical-specs":{"camera": "Arriflex 35-III"}}]
```

Perhaps the `%20` is being taken literally (as-is) in the tool you're using? Maybe try writing the space directly `eq.Arriflex 35-III`.

**Reply by christopheraranda** [reply]
I'm testing using swagger-ui which seems to be encoding the `=` for the camera filter from
`technical-specs.camera=eq.Arriflex%2035-III`
to
`technical-specs.camera%3Deq.Arriflex%2035-III`
I'm not sure if that would make a difference?

**Reply by wolfgangwalther** [reply]
> I'm not sure if that would make a difference?

Yeah, that's a big difference, imho.

**Reply by laurenceisla** [reply, marked as answer]
> Yeah, that's a big difference, imho.

Ah, yeah, that changes everything. The Swagger output does not allow for custom query parameters right now, so the `technical-specs.camera` parameter cannot be added to the `/films` endpoint. It's a limitation for now.

I'll open an issue to keep track of it. This will be addressed in the [`postgrest-openapi` project](https://github.com/PostgREST/postgrest-openapi).

Edit: Opened https://github.com/PostgREST/postgrest/issues/3217

**Reply by christopheraranda** [reply]
@laurenceisla & @wolfgangwalther I see. Thanks so much for your help! I did notice that the spec being produced was Swagger 2.0 and not OpenAPI 3.x. I didn't realize that would make a difference. I'll switch to using curl or some other tool for testing. Thanks again!
